airflow/jobs.py

```diff
@@ -213,27 +212,54 @@ def _execute(self):
         raise NotImplementedError("This method needs to be overridden")
 
     @provide_session
-    def reset_state_for_orphaned_tasks(self, dag_run, session=None):
+    def reset_state_for_orphaned_tasks(self, filter_by_dag_run=None, session=None):
         """
-        This function checks for a DagRun if there are any tasks
+        This function checks if there are any tasks in the dagrun (or all)
         that have a scheduled state but are not known by the
         executor. If it finds those it will reset the state to None
         so they will get picked up again.
+        The batch option is for performance reasons as the queries are made in
+        sequence.
+
+        :param filter_by_dag_run: the dag_run we want to process, None if all
+        :type filter_by_dag_run: models.DagRun
+        :return: the number of reset TIs
+        :rtype: int
         """
         queued_tis = self.executor.queued_tasks
 
         # also consider running as the state might not have changed in the db yet
         running = self.executor.running
-        tis = list()
-        tis.extend(dag_run.get_task_instances(state=State.SCHEDULED, session=session))
-        tis.extend(dag_run.get_task_instances(state=State.QUEUED, session=session))
-
-        for ti in tis:
+        resettable_states = [State.SCHEDULED, State.QUEUED]
+        TI = models.TaskInstance
+        DR = models.DagRun
+        if filter_by_dag_run is None:
+            tis_to_reset = (session.query(TI)
+                .join(DR,
+                      and_(
+                          TI.dag_id == DR.dag_id,
+                          TI.execution_date == DR.execution_date))
+                .filter(
+                    DR.state == State.RUNNING,
+                    DR.external_trigger.is_(False),
+                    DR.run_id.notlike('%backfill%'),
+                    TI.state.in_(resettable_states))).all()
+        else:
+            tis_to_reset = filter_by_dag_run.get_task_instances(state=resettable_states,
+                                                                session=session)
+        tis_reset = []
+        # Sort by dag_id, task_id, execution_date to prevent deadlock
+        # Can't use an update here since it doesn't support joins
+        tis_to_reset.sort(key=lambda x: (x.dag_id, x.task_id, x.execution_date))
+        for ti in tis_to_reset:
             if ti.key not in queued_tis and ti.key not in running:
                 self.logger.debug("Rescheduling orphaned task {}".format(ti))
```

Comment:
This is only slightly better. Before we did one query for each active dagrun, then one update for each TI. Now there's just 1 + 2 * TI.